README.md
Oblong Bootstrap Script
=======================

`obs` is a tiny tool for mac and windows that can install local packages which don't belong in the app store or brew.

It does not support transitive dependencies -- it just finds the tarball and unpacks it.

(On Linux, use the system's package manager instead.)

## Getting Started

First, arrange for passwordless login to the package server using your LDAP credentials, e.g.:

```
$ ssh-copy-id buildhost5.oblong.com
```

Then:

On Mac, [install brew if you don't already have it](https://brew.sh/), then do:

```
$ brew tap Oblong/tools
$ brew install obs
```

On Ubuntu, do:

```
$ sudo apt-get install oblong-obs
```

On Windows, in a Cygwin window, do:

```
$ git clone https://github.com/Oblong/obs.git
$ cd obs
$ make install
```

## Usage

### Installing packages

```
$ obs install yobuild11 g-speak3.31
```

### Listing available packages

```
$ obs list
```

You can search for particular packages by passing the output of 'obs list' to a tool like [grep](http://www.uccs.edu/~ahitchco/grep/).

## Troubleshooting

If obs acts funny, see if the latest version works better, e.g. on Mac, do

```
$ brew upgrade obs
```

If cmake explodes, it probably means you missed installing a package with obs.   You can guess what's needed by looking at debian/control and/or cmake errors, then figure out the name of the exact package to install with `obs list`.

If programs crash, you're probably mixing old and new packages.

To fix that for packages installed with obs, your best bet is to remove all packages and start over with the rather extreme:

```
$ sudo rm -rf /opt/oblong
```

(On Cygwin, it's ```rm -rf /cygdrive/c/opt/oblong```.)

## Using list and install together

For example, to install greenhouse and webthing, use 'obs list' to discover dependencies, then 'obs install' them:

```
$ obs list | grep yobuild
oblong-yobuild11-cef2704
oblong-yobuild11-cef3112
oblong-yobuild11-v8-4.2.77
oblong-yobuild11-v8-5.2.361
oblong-yobuild12-cef3112
yobuild11
yobuild12
$ obs install yobuild11 oblong-yobuild11-cef2704
$ obs list | egrep 'g-speak|webthing|greenhouse|staging' | grep 3.31
g-speak3.31
g-speak-gh3.31
oblong-greenhouse-gs3.31x
oblong-greenhouse-gs-gh3.31x
oblong-webthing-cef2704-gs3.31x
oblong-webthing-cef2704-gs-gh3.31x
staging3.31
staging-gh3.31
$ obs install g-speak3.31 oblong-greenhouse-gs3.31x oblong-webthing-cef2704-gs3.31x staging3.31
```

## Other functions and options

`obs --help` lists other things obs can do.  For instance:

To look up the version of yobuild that a given release of the oblong platform uses:

```
$ obs yovo2yoversion 3.20
10
```

To look up the version of cef that a given release of the oblong platform uses:

```
$ obs yovo2cefversion 3.20
cef2272
```

To install some other operating system's build of a package
(handy e.g. on osx1013 if all that's been built is for osx1012):

```
$ BS_FORCE_OS=osx1012 obs install yobuild11 g-speak3.28
```

To use a different username when accessing the package server:

```
$ bs_install_user=joe obs list
```

Alternately, add something like this to ~/.ssh/config:

```
Host buildhost5.oblong.com
  User joe
```

To access a different package server:

```
$ MASTER=server.example.com obs pkg_list
```

## ob-set-defaults

ob-set-defaults modifies a project's source tree slightly to set defaults
for e.g. which version of g-speak to install when building the project.

Individual projects can override this by putting their own version
in an executable sh script named set-gspeak.sh.

More simply, they can override some defaults by creating a file
ci/ob-set-defaults.conf containing variable settings like
```
PREFIX=/opt/oblong/my-funky-dir
opt_generator="Unix Makefiles"
```
Read the top of ob-set-defaults to see which variables can be set.

See ob-set-defaults --help for more info.


debian/changelog
oblong-obs (1.33) bionic; urgency=low

  * Support yobuild13.

 -- Dan Kegel <dank@oblong.com>  Tue, 23 Apr 2019 13:59:24 -0700

oblong-obs (1.32) bionic; urgency=low

  * No user visible changes, tagging a release to test the pipeline.

 -- Dan Kegel <dank@oblong.com>  Tue, 02 Apr 2019 13:01:36 -0700

oblong-obs (1.31) bionic; urgency=low

  * Makefile: don't test scancode until it can be installed reliably via pip

 -- Dan Kegel <dank@oblong.com>  Sat, 30 Mar 2019 22:02:17 -0700

oblong-obs (1.30) bionic; urgency=low

  * uberbau: on ubu1604, mirror guile, too
  * uberbau: on ubu1804, mirror nodejs6, too
  * debian: depend on python-wheel so scancode can install

 -- Dan Kegel <dank@oblong.com>  Sat, 30 Mar 2019 11:48:35 -0700

oblong-obs (1.29) bionic; urgency=low

  * obs, bau: default to buildhost5
  * ob-list-licenses: new tool
  * master.json: webthing-cef: removed 3396, added 3497, 3578
  * uberbau: bugfixes
  * better support for msvc2017

 -- Dan Kegel <dank@oblong.com>  Fri, 08 Feb 2019 15:05:20 -0800

oblong-obs (1.23) bionic; urgency=low

  * ob-set-defaults: when setting g-speak, get default yobuild from
    g-speak version rather than project.
  * ob-set-defaults: run ci/ob-set-defaults.extra if present
  * uberbau: mirror nobuild: remove obsolete qt 5.3 and 5.5, document and tidy 5.9
  * bau-defaults: avoid ambiguity about whether script should abort if xset missing
  * bau: abort if mixing two incompatible sources of packages
  * run-bslave-in-terminal: new tack on avoiding terminal window pile-up on mac
  * buildbot/master.cfg: don't build all commits of a merge

 -- Dan Kegel <dank@oblong.com>  Fri, 03 Aug 2018 11:36:56 -0700

oblong-obs (1.22) bionic; urgency=low

  * uberbau: mirror third-party rather than nodejs6; check YOVERSION, too.
  * obs.bats: don't fail if g-speak not yet built
  * uberbau: no longer mirror qt531 and westar; they are no longer used as of g-speak 4.2
  * uberbau: mirror: handle buildhost5 builder names better
  * uberbau: adjust to buildhost5; also, defer mirroring obs to avoid bootstrap failure.
  * bau: on ubuntu, keep obs in front of path; fixes obs build error due to stale obs.
  * Obey MASTER more thoroughly.
  * bslave.sh: fix usage message, handle missing pip3 on ubuntu
  * Makefile: get manpage to build on osx
  * buildbot/run-bslave: call bootstrap-obs.sh at startup
  * bau-defaults/buildshim-ubu: clear dependencies if not in container
  * buildbot/run-bslave: start x if it's not already started on linux
  * buildbot/master.cfg: fix intuit_buildtype typo, add git describe step,
    shorten name of builders in gitlab ui

 -- Dan Kegel <dank@oblong.com> Thu, 19 Jul 2018 14:18:00 -0700

oblong-obs (1.21) bionic; urgency=low

  * buildbot/bslave.sh: support ubu1604
  * buildbot: helper scripts for running workers on mac, etc.
  * buildbot/bslave.sh: first draft of script to show how to create a worker.
  * buildbot/master.cfg: don't override default for gitlab status reporting; a static string isn't enough.
  * bs_is_try_build: also recognize pwd of *-try-* as a try build
  * buildbot/bslave.sh: Try to leave virtualenv in control; see platform/ob-repobot#76
  * bs_install_xcode_version: if xcodebuild aborts, run xcode-select unconditionally.
  * bau-defaults/buildshim-ubu: the pre-purging kludge makes using bau
    hard, so only do it if user is buildbot.

 -- Dan Kegel <dank@oblong.com> Tue, 10 Jul 2018 11:15:20 -0700

oblong-obs (1.20) bionic; urgency=low

  * bau: make -h show project-specific help, too
  * bau-defaults/buildshim-ubu: add --headless option to unset DISPLAY
  * obs: support osx1014
  * obs: Add purge-build (to delete build directory and kill straggling jobs)
  * Fix greenhouse tdt on ubu1804

 -- Dan Kegel <dank@oblong.com>  Tue, 19 Jun 2018 11:12:22 -0700

oblong-obs (1.19) bionic; urgency=low

  * On Ubuntu, default buildshim now handles --asan
    and uses ci/{tsan,lsan}-{suppressions,options}.txt if they exist.
    This makes it a lot easier to use asan.
  * When --asan or --tsan is given, add -as to package names.
    This makes it a lot easier to get and provide prebuilt sanitized
    dependencies.
  * Known issue: our packages' various flavors
    (e.g. -as and non-as, -cef3112 and -cef3282)
    install to the same location, but don't yet tell apt they replace
    each other, so apt errors out if you install one on top of the other.
  * The above is particularly painful on non-containerized
    ubuntu buildbots (and our 18.04 ones aren't containerized yet).
  * As a temporary expedient, on ubuntu 18.04, 'bau install-deps'
    does 'ob-remove.sh' first for now (!).
  * ob-set-defaults now handles oblong-rtsp-viddle-server properly

 -- Dan Kegel <dank@oblong.com>  Wed, 16 May 2018 13:02:21 -0700

oblong-obs (1.18) precise; urgency=low

  * Fix uberbau on mac without greadlink

 -- Dan Kegel <dank@oblong.com>  Fri, 20 Apr 2018 10:51:12 -0700

oblong-obs (1.17) precise; urgency=low

  * Fix trickle-down builds

 -- Dan Kegel <dank@oblong.com>  Tue, 17 Apr 2018 08:13:00 -0700

oblong-obs (1.15) precise; urgency=low

  * Support ubuntu 18.04

 -- Dan Kegel <dank@oblong.com>  Thu, 12 Apr 2018 07:12:48 -0700

oblong-obs (0.1) precise; urgency=low

  * Initial packaging

 -- Dan Kegel <dank@oblong.com>  Mon, 04 Jan 2016 13:00:40 -0800


buildbot/purge.sh
#!/bin/sh
# Clean up large download artifacts (not in apt) older than one month
# Note: if disk fills up, you may have to delete zero-sized .pid files
# before buildbot will start up again.

# See also version-cleaner.sh, for those one-time cleanups of old versions.

set -x

RECIPIENTS="dank@oblong.com"
DIRS="/buildhost5_archive /home /"

date
df $DIRS

# Delete old versions.  We haven't been doing this, but it might save lots of space.
#timeout 120m ~/src/ob-repobot/arepo.sh flense

#--- 
# This does a full find on ~/.aptly, which takes AGES.
# Doing it in purge.sh save time because then it can share cache with
# the du.
# See also https://gitlab.oblong.com/platform/ob-repobot/issues/39
#timeout 120m ~/src/ob-repobot/arepo.sh cleanup

#--- Note disk usage ---
oldlogdir=`ls -trd ~/dulogs/* | tail -n 1`
datecode=`date +%F-%R`
newlogdir=$HOME/dulogs/$datecode
mkdir $newlogdir
sudo du --one-file-system $DIRS > $newlogdir/du.raw
sort -k +2 < $newlogdir/du.raw > $newlogdir/du.log
join -j 2 $oldlogdir/du.log $newlogdir/du.log | awk '$3 > $2 {print $3-$2, $1}' | sort -n -k +3 | sort -rn > $newlogdir/growth.log


#--- clean up artifact repositories ---
# First with a wide search and a lenient timeframe, then with a targeted search and a strict timeframe
# lenient = 30 days
# dev_lifetime = 14 days (and we should do a weekly build of those bits so they stay fresh)
# try_lifetime = 2 days
soft_lifetime=25
dev_lifetime=13
dev_lifetime_short=6
try_lifetime=4    # allow friday tries to live until end of monday

#--- purge trybuilders ---
find -L /var/repobot/*/builds/*-trybuilder* -type f -mtime +$try_lifetime | xargs --no-run-if-empty rm -vf

#--- purge repobot/tarballs ---
find -L /var/repobot/tarballs -type f -ctime +$soft_lifetime | egrep '\.deb|\.dmg|\.pkg' | xargs  --no-run-if-empty rm -vf
known_offenders="
 oblong-*-gs3.*[13579]x
 mezzanine-*/*/*[13579]
 staging3.*[13579]
 staging-gh3.*[13579]
 oblong-platform/*/*[13579]
"
# This causes awful horrible problem: old packages being picked up as new
#(cd /var/repobot/tarballs; find -L $known_offenders -type f -ctime +$dev_lifetime -size +100k | xargs  --no-run-if-empty rm -vf)
df /var/repobot

#--- purge repobot/dev/builds ---
# First line purges all g-speak 3.19 build products; useful only very briefly after 3.21 development starts
#find /var/repobot/dev/builds -type f -ctime +$soft_lifetime | egrep 'gs3\.19|3\.19-3\.19' | xargs  --no-run-if-empty rm -rvf
find -L /var/repobot/dev/builds -type f -ctime +$soft_lifetime | egrep '\.deb|\.dmg|\.pkg' | xargs  --no-run-if-empty rm -vf
find -L /var/repobot/dev/builds -type d -ctime +$soft_lifetime | egrep 'coverity.*output' | xargs  --no-run-if-empty rm -rvf
known_offenders="
 mezzanine-*
 growroom-*
 platform-*
"
(cd /var/repobot/dev/builds; find -L $known_offenders -type f -ctime +$dev_lifetime -size +100k | xargs  --no-run-if-empty rm -vf)
df /var/repobot

#--- purge repobot/dev/builds with shorter lifespan for builds with large artifacts ---
known_offenders="
 appup-generation-*
"
(cd /var/repobot/dev/builds; find -L $known_offenders -type f -ctime +$dev_lifetime_short -size +100k | xargs  --no-run-if-empty rm -vf)
df /var/repobot

#--- purge repobot/dev/builds with 2 days lifespan for builds whose artifacts aren't crucial ---
known_offenders="
 mezzanine-ubu1204gl-mz
 mezzanine-ubu1204gl-mz-coverage
 meow-*-longtests
"
(cd /var/repobot/dev/builds; find -L $known_offenders -type f -ctime +2 -size +10k | xargs  --no-run-if-empty rm -vf)
df /var/repobot
(cd /var/repobot/rel/builds; find -L $known_offenders -type f -ctime +2 -size +10k | xargs  --no-run-if-empty rm -vf)
df /var/repobot

#--- purge repobot/rel/builds ---
known_offenders="
 *-master3.*[13579]
 *-master3.*[13579]-gh
 *-master3.*[13579]-msvc*
"
(cd /home/buildbot/var/repobot/rel/builds; find -L $known_offenders -type f -ctime +7 -size +8k | xargs  --no-run-if-empty rm -vf)
df /var/repobot

known_offenders="
 mezzanine-ubu1204-rel-3.0-mz-chiltepe
 mezzanine-ubu1404-rel-3.0-mz-chiltepe
"
(cd /home/buildbot/var/repobot/rel/builds; find -L $known_offenders -type f -ctime +$soft_lifetime -size +8k | xargs  --no-run-if-empty rm -vf)
df /var/repobot

#--- purge package-monster ---
# big offender is platform-builder
# astor only needs 3.16 and 3.20, and only on ubu1204
ssh package-monster.oblong.com rm -rf /ob/dumper/g-speak/platform/ubu1[04]04 /ob/dumper/g-speak/platform/ubu1204/*gs3.1[02489]*gz

#--- purge obdumper ---
ssh git.oblong.com "df /ob/dumper; find /ob/dumper/g-speak/platform -name 'oblong-*-gs[3-9]\.*[13579]-*-*' -ctime +$dev_lifetime | xargs  --no-run-if-empty rm -vf; df /ob/dumper"

#--- report ---

> report.txt
report_queue()
{
    echo "$@" >> report.txt
}
report_by_os()
{
    (
    cd /var/repobot
    echo ===========================
    echo "Total usage on master by operating system:"
    echo -n ubu1004 " "; du *lucid*   */builds/*1004* -chs | tail -n 1
    echo -n ubu1204 " "; du *precise* */builds/*1204* -chs | tail -n 1
    echo -n ubu1404 " "; du *trusty*  */builds/*1404* -chs | tail -n 1
    echo -n ubu1510 " "; du *wily*    */builds/*1510* -chs | tail -n 1
    echo -n ubu1604 " "; du *xenial*  */builds/*1604* -chs | tail -n 1
    echo -n ubu1704 " "; du *zesty*   */builds/*1704* -chs | tail -n 1
    echo -n ubu1710 " "; du *artful*  */builds/*1710* -chs | tail -n 1
    echo -n osx107  " "; du tarballs/*/*osx107* */builds/*osx107* -chs | tail -n 1
    echo -n osx109  " "; du tarballs/*/*osx109* */builds/*osx109* -chs | tail -n 1
    echo -n osx1010 " "; du tarballs/*/*osx1010* */builds/*osx1010* -chs | tail -n 1
    echo -n osx1011 " "; du tarballs/*/*osx1011* */builds/*osx1011* -chs | tail -n 1
    echo -n osx1012 " "; du tarballs/*/*osx1012* */builds/*osx1012* -chs | tail -n 1

    )
}

report_send()
{
    if test -s report.txt
    then
        #report_by_os >> report.txt
        mail $RECIPIENTS -s "bot disk space warning" < report.txt
    fi
    rm report.txt
}

for slave in `cat ~/src/ob-repobot/slaves.txt`
do
    used=`ssh buildbot@${slave} df \~buildbot/slave-state | grep / | grep '%' | sed 's/%.*//;s/.* //'`
    if test "$used" -gt 80
    then
        report_queue "WARNING: $slave disk space warning: $used% used"
    fi
    used=`ssh buildbot@${slave} df / | grep / | grep '%' | sed 's/%.*//;s/.* //'`
    if test "$used" -gt 80
    then
        report_queue "WARNING: $slave root disk space warning: $used% used"
    fi
done

for slave in git.oblong.com package-monster.oblong.com
do
    used=`ssh buildbot@${slave} df /ob/dumper | egrep / | grep '%' | sed 's/%.*//;s/.* //'`
    if test "$used" -gt 80
    then
        report_queue "WARNING: $slave disk space warning: $used% used"
    fi
done

# Local problems.  Has to be run on the buildhost.
problem=false
for dir in / /home /superfast_archive
do
    used=`df $dir | grep / | grep '%' | sed 's/%.*//;s/.* //'`
    if test "$used" -gt 75
    then
        report_queue "WARNING: `hostname` disk space warning: $dir is $used% used"
        problem=true
    fi
done
if $problem
then
    report_queue "$newlogdir/growth.log shows growth in these directories:"
    head -n 40 < $newlogdir/growth.log >> report.txt
fi

# Monitor for recurring vmware configuration snafu
# Assumes ob-version is installed everywhere, maybe we should put a
# special copy of it somewhere so it doesn't accidentally get uninstalled
for slave in `cat ~/src/ob-repobot/slaves.txt | grep -v win`
do
    cpu1=`ssh buildbot@${slave} grep -i model.name /proc/cpuinfo | uniq | sed 's/.*://'`
    cpu2=`ssh buildbot@${slave} ob-version | grep cpu.version | sed 's/.*://'`
    if test "$cpu1" != "$cpu2"
    then
        report_queue "$slave configuration warning: cpu names don't match ('$cpu1' != '$cpu2'), see bug 14946"
    fi
done

report_send

